include/irr/asset/material_compiler/CMaterialCompilerGLSLBackendCommon.h

```diff
@@ -0,0 +1,506 @@
+	_IRR_STATIC_INLINE_CONSTEXPR uint32_t BITFIELDS_FETCH_TEX_2_REG_CNT_SHIFT = BITFIELDS_FETCH_TEX_1_REG_CNT_SHIFT + BITFIELDS_FETCH_TEX_REG_CNT_WIDTH;
+	_IRR_STATIC_INLINE_CONSTEXPR uint32_t BITFIELDS_FETCH_TEX_3_REG_CNT_SHIFT = BITFIELDS_FETCH_TEX_2_REG_CNT_SHIFT + BITFIELDS_FETCH_TEX_REG_CNT_WIDTH;
+	static_assert(BITFIELDS_FETCH_TEX_3_REG_CNT_SHIFT+BITFIELDS_FETCH_TEX_REG_CNT_WIDTH <= 8u*sizeof(instr_t));
+
+	inline uint32_t getTexFetchFlags(instr_t i)
+	{
+		auto flags = core::bitfieldExtract(i, BITFIELDS_FETCH_TEX_0_SHIFT, 1);
+		flags |= core::bitfieldExtract(i, BITFIELDS_FETCH_TEX_1_SHIFT, 1)<<1;
+		flags |= core::bitfieldExtract(i, BITFIELDS_FETCH_TEX_2_SHIFT, 1)<<2;
+		flags |= core::bitfieldExtract(i, BITFIELDS_FETCH_TEX_3_SHIFT, 1)<<3;
+
+		return static_cast<uint32_t>(flags);
+	}
+}
+namespace normal_precomp
+{
+	_IRR_STATIC_INLINE_CONSTEXPR uint32_t BITFIELDS_REG_DST_SHIFT = remainder_and_pdf::INSTR_REG_DST_SHIFT;
+	_IRR_STATIC_INLINE_CONSTEXPR uint32_t BITFIELDS_REG_WIDTH = remainder_and_pdf::INSTR_REG_WIDTH;
+}
+}
+
+class CMaterialCompilerGLSLBackendCommon
+{
+public:
+	struct result_t;
+	struct SContext;
+
+protected:
+	_IRR_STATIC_INLINE_CONSTEXPR const char* OPCODE_NAMES[instr_stream::OPCODE_COUNT]{
+		"OP_DIFFUSE",
+		"OP_CONDUCTOR",
+		"OP_PLASTIC",
+		"OP_COATING",
+		"OP_DIFFTRANS",
+		"OP_DIELECTRIC",
+		"OP_THINDIELECTRIC",
+		"OP_BLEND",
+		"OP_BUMPMAP",
+		"OP_SET_GEOM_NORMAL",
+		"OP_INVALID",
+		"OP_NOOP"
+	};
+	_IRR_STATIC_INLINE_CONSTEXPR const char* NDF_NAMES[instr_stream::NDF_COUNT]{
+		"NDF_BECKMANN",
+		"NDF_GGX",
+		"NDF_PHONG"
+	};
+	static std::string genPreprocDefinitions(const result_t& _res, bool _genChoiceStream);
+
+	core::unordered_map<uint32_t, uint32_t> createBsdfDataIndexMapForPrefetchedTextures(SContext* _ctx, const instr_stream::traversal_t& _tex_prefetch_stream, const core::unordered_map<instr_stream::STextureData, uint32_t, instr_stream::STextureData::hash>& _tex2reg) const;
+
+	void adjustBSDFDataIndices(instr_stream::traversal_t& _stream, const core::unordered_map<uint32_t, uint32_t>& _ix2ix) const
+	{
+		using namespace instr_stream;
+		for (instr_t& i : _stream) {
+			auto found = _ix2ix.find(getBSDFDataIx(i));
+			if (found != _ix2ix.end())
+				setBSDFDataIx(i, found->second);
+		}
+	}
+	void setSourceRegForBumpmaps(instr_stream::traversal_t& _stream, uint32_t _regNumOffset)
+	{
+		using namespace instr_stream;
+		for (instr_t& i : _stream) {
+			if (getOpcode(i)==OP_BUMPMAP)
+			{
+				const uint32_t n_id = getNormalId(i);
+				i = core::bitfieldInsert<instr_t>(i, _regNumOffset+n_id, INSTR_BUMPMAP_SRC_REG_SHIFT, INSTR_BUMPMAP_SRC_REG_WIDTH);
+			}
+		}
+	}
+
+	//common for rem_and_pdf and gen_choice instructions
+	void debugPrintInstr(std::ostream& _out, instr_stream::instr_t instr, const result_t& _res, const SContext* _ctx) const;
+
+public:
+	struct SContext
+	{
+		//users should not touch this
```

Comment:
maybe make everything `private` or `protected` except for the VT and then make the parent class a `friend`

Reply:
done?

Reply:
yes